File: packages/python/python-shared/src/python_shared/http/pipeline/step_rate_limit.py
from python_shared.http.pipeline.types import PipelineContext

class StepRateLimit:
    name: str = "RateLimit"
    description: str = "Per-Tenant Token Bucket Outbound Rate Limiting"

    async def execute(self, ctx: PipelineContext) -> None:
        ctx.step_index += 1
        if not ctx.rate_limiter.allow_request(ctx.tenant_id):
            raise RuntimeError(f"Rate limit exceeded for tenant {ctx.tenant_id}")
        ctx.retry_budget.record_request()

File: packages/llm-obs-node-packages/shared-infra/src/http/pipeline/step-rate-limit.ts
/**
 * @file step-rate-limit.ts
 * @description Pipeline Step 4: Token Bucket Rate Limiting per Tenant.
 */

import type { PipelineStep, PipelineContext } from "./types";

export class StepRateLimit implements PipelineStep {
  public readonly name = "RateLimit";
  public readonly description = "Per-Tenant Token Bucket Outbound Rate Limiting";

  public async execute<T>(ctx: PipelineContext<T>): Promise<void> {
    ctx.stepIndex++;
    if (!ctx.rateLimiter.allowRequest(ctx.tenantId)) {
      throw new Error(`Rate limit exceeded for tenant ${ctx.tenantId}`);
    }
    ctx.retryBudget.recordRequest();
  }
}
